Sources/SpeziFirestore/DocumentReference+AsyncAwait.swift

```diff
@@ -6,13 +6,128 @@
 // SPDX-License-Identifier: MIT
 //
 
-
-@_exported import FirebaseFirestore
-@_exported import FirebaseFirestoreSwift
+import FirebaseFirestore
 import Foundation
 
 
 extension DocumentReference {
+#if compiler(>=6)
+    /// Overwrite the data of a document with an encodable value.
+    ///
+    /// Encodes an instance of `Encodable` and overwrites the encoded data
+    /// to the document referred by this `DocumentReference`. If no document exists,
+    /// it is created. If a document already exists, it is overwritten.
+    ///
+    /// See `Firestore.Encoder` for more details about the encoding process.
+    ///
+    /// Returns once the document has been successfully written to the server.
+    /// Due to the Firebase SDK, it will not return when the client is offline, though local changes will be visible immediately.
+    ///
+    /// - Parameters:
+    ///   - value: An instance of `Encodable` to be encoded to a document.
+    ///   - encoder: An encoder instance to use to run the encoding.
+    public func setData<T: Encodable>( // swiftlint:disable:this function_default_parameter_at_end
+        isolation: isolated (any Actor)? = #isolation,
+        from value: T,
+        encoder: FirebaseFirestore.Firestore.Encoder = FirebaseFirestore.Firestore.Encoder()
+    ) async throws {
+        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
```

Comment:
We have made strange experiences with the async behavior of Firebase to call the completion handler twice and therefore moved to a very simplified setup in #38. Just want to make sure that we avoid a similar element here and ideally use the existing async overloads and map to them as much as possible if they existing.

Reply:
We sadly have never investigated the issue fully (see [comment](https://github.com/StanfordSpezi/SpeziFirebase/pull/38#issuecomment-2285675976)). Was it verified that the error was due to the completion handler getting called twice? Any indication that this was fixed in the latest Firestore SDK?

For context, I chose this design as it avoids a concurrency warning that pops up with Strict Concurrency checking.
The problem with the existing implementation (which is still there for compatibility with Swift 5.10 toolchains), is that it is not concurrency safe. Encoding happens on an async queue and might introduce race conditions. Therefore, the methods inherit the actor isolation from their called if the Swift 6 toolchain is available. However, when using the native `setData(_:encoder:)` methods provided by Firestore (which are Objective C methods with completion handlers that get automatically converted to Swift-concurrency methods) I get a warning like "Sending 'encoded._bridgeToObjectiveC' risks causing data races;" as the Firestore methods expect a NSDictionary and something with the automatic bridging to Objective C seems unsafe from a concurrency perspective.

Reply:
Thank you for the context @Supereg!

The approach that we are re-introducing here [is the same that we had before](https://github.com/StanfordSpezi/SpeziFirebase/pull/38/files#diff-207c9bfc6818a064bbfb0d2f9e5a4c37ec8144743cd9eef0c5201f4f2291818eL33) and I am a bit worried that it will lead to the same exact runtime behavior.

To be honest: At the end we gave up and just switched to the Firebase Async methods and it worked fine since then. Not sure if Firebase has fixed that regression and we came to the conclusion that it was either the continuation being called multiple times or an unsafe memory access in there due to a race condition or similar.

The general question is if we want to solve this in here or if there is a more reasonable approach to contribute this to the main Firebase SDK and ensure that maintenance and in-depth testing is happening there. IMO these are small oversights in the SDK that are never addressed as not a lot of people might use the async functions as of now and Firebase might not be aware of the latest changes and advancements in Swift 6?

Reply:
I wonder how the "native" methods do not expose the same error. The completion handler based methods get automatically translated to async/await methods, see [SE-0297 Concurrency Interoperability with Objective-C](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0297-concurrency-objc.md). Under the hood it should basically translate to the same code. The difference is that they use `withUnsafeContinuation`. Both specify that calling resume multiple times is undefined behavior, however not sure if an `UnsafeContinuation` crashes when resuming multiple times. I will test.
But likely, the issue is present in both implementations, it might just not surface when using the automatic interoperability.

In general, if we find an underlying issue in Firebase it makes sense to contribute back. I'm just not sure how easily reproducible it is or if it might have been already fixed. Contribution back Swift 6 related improvements seems hard as I don't quite know how to optimize for Swift Concurrency within Objective C code. I'll review some of the latest release notes to see if there is anything indicating something like this was fixed.

Reply:
I have disproven my theory. Both the `UnsafeContinuation` and the `CheckedContinuation` crash when trying to be resumed multiple times. It's just that `CheckedContinuation` adds more debug information.

The proposal seems to be aware of this potential issue, see [Completion handlers must be called exactly once](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0297-concurrency-objc.md#completion-handlers-must-be-called-exactly-once). However, it is not quite clear what the synthesized completion block does in such a scenario. A question about this in the forums seems to have gone unanswered https://forums.swift.org/t/accepted-with-modifications-se-0297-concurrency-interoperability-with-objective-c/43306/10

Reply:
Interesting; thank you for the further inspections @Supereg!

Wondering if we should add some additional state handling here that ensures continuations to only be called once and ignore any further calls from Firebase or other places to harden this a bit more. If we have that in place I am fine with trying to get this merged & seeing if we encounter any further issues down the road.

Reply:
Added a lock-free safety guard around completion handlers.

My further research into the Firebase iOS SDK didn't provide any further insights. Every setData call end up calling [SyncEngine:WriteMutations(...)(https://github.com/firebase/firebase-ios-sdk/blob/9118aca998dbe2ceac45d64b21a91c6376928df7/Firestore/core/src/core/sync_engine.cc#L235-L245) which is part of the C++ library. The method gets passed a [ParsedUpdateData](https://github.com/firebase/firebase-ios-sdk/blob/9118aca998dbe2ceac45d64b21a91c6376928df7/Firestore/Source/API/FIRDocumentReference.mm#L185)  instance. I didn't invest more time to research all code paths that might lead to completion of a write mutation.